File: practice-projects/gin-ddd/internal/interfaces/vo/user/user_response.go
package user

import (
	userDTO "gin-ddd/internal/application/dto/user"
	"time"
)

// UserResponse 用户响应
type UserResponse struct {
	ID        int64     `json:"id"`
	Name      string    `json:"name"`
	Email     string    `json:"email"`
	Phone     *string   `json:"phone,omitempty"`
	CreatedTime time.Time `json:"created_time"`
	UpdatedTime time.Time `json:"updated_time"`
}

// FromUserDTO 将用户DTO转换为响应VO
func FromUserDTO(dto *userDTO.UserDTO) *UserResponse {
	if dto == nil {
		return nil
	}
	return &UserResponse{
		ID:          dto.ID,
		Name:        dto.Name,
		Email:       dto.Email,
		Phone:       dto.Phone,
		CreatedTime: dto.CreatedTime,
		UpdatedTime: dto.UpdatedTime,
	}
}

// FromUserDTOs 批量将用户DTO转换为响应VO
func FromUserDTOs(dtos []*userDTO.UserDTO) []*UserResponse {
	responses := make([]*UserResponse, 0, len(dtos))
	for _, dto := range dtos {
		responses = append(responses, FromUserDTO(dto))
	}
	return responses
}

File: practice-projects/nestjs-ddd/src/interfaces/http/user/vo/user.response.ts
import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger';
import { UserDto } from '../../../../application/user/dto/user.dto';

/**
 * 用户接口响应 VO - JSON 字段采用 snake_case，与 gin-ddd 对齐。
 */
export class UserResponse {
  @ApiProperty()
  id!: number;

  @ApiProperty()
  name!: string;

  @ApiProperty()
  email!: string;

  @ApiPropertyOptional()
  phone?: string;

  @ApiPropertyOptional()
  address?: string;

  @ApiProperty({ name: 'created_time' })
  created_time!: Date;

  @ApiProperty({ name: 'updated_time' })
  updated_time!: Date;

  static fromDto(dto: UserDto): UserResponse {
    const vo = new UserResponse();
    vo.id = dto.id;
    vo.name = dto.name;
    vo.email = dto.email;
    vo.phone = dto.phone;
    vo.address = dto.address;
    vo.created_time = dto.createdTime;
    vo.updated_time = dto.updatedTime;
    return vo;
  }

  static fromDtos(dtos: UserDto[]): UserResponse[] {
    return dtos.map((d) => UserResponse.fromDto(d));
  }
}
